File: tests/python/test_field.py
# Gingo — Field tests (Python)
# SPDX-License-Identifier: MIT

import pytest
from gingo import Field, ScaleType, Chord, HarmonicFunction


def test_field_construction():
    f = Field("C", ScaleType.Major)
    assert f.size() >= 7


def test_field_triads():
    f = Field("C", ScaleType.Major)
    triads = f.chords()
    assert len(triads) >= 7
    # All returned elements should be Chord objects
    assert all(isinstance(c, Chord) for c in triads)


def test_field_sevenths():
    f = Field("C", ScaleType.Major)
    sevenths = f.sevenths()
    assert len(sevenths) >= 7
    assert all(isinstance(c, Chord) for c in sevenths)


def test_field_chord():
    f = Field("C", ScaleType.Major)
    first = f.chord(1)
    assert isinstance(first, Chord)


def test_field_chord_out_of_range():
    f = Field("C", ScaleType.Major)
    with pytest.raises(IndexError):
        f.chord(0)
    with pytest.raises(IndexError):
        f.chord(100)


def test_field_string_name():
    f = Field("C", ScaleType.Major)
    assert "Field" in repr(f) or "C" in str(f)


# ---------------------------------------------------------------------------
# Applied chords (tonicization)
# ---------------------------------------------------------------------------

def test_applied_string_int():
    f = Field("C", ScaleType.Major)
    # V7/II = A7
    c = f.applied("V7", 2)
    assert c.root().natural() == "A"
    assert c.type() == "7"
    # V7/IV = C7
    c = f.applied("V7", 4)
    assert c.root().natural() == "C"
    assert c.type() == "7"
    # V7/V = D7
    c = f.applied("V7", 5)
    assert c.root().natural() == "D"
    assert c.type() == "7"


def test_applied_complex_quality():
    f = Field("C", ScaleType.Major)
    # IIm7(b5)/V = Am7(b5)
    c = f.applied("IIm7(b5)", 5)
    assert c.root().natural() == "A"
    assert c.type() == "m7(b5)"


def test_applied_no_quality():
    f = Field("C", ScaleType.Major)
    # V/II = triad at V of D major = A
    c = f.applied("V", 2)
    assert c.root().natural() == "A"


def test_applied_accidental():
    f = Field("C", ScaleType.Major)
    # bVII7/V = F7
    c = f.applied("bVII7", 5)
    assert c.root().natural() == "F"
    assert c.type() == "7"
    # #IV7/I = F#7
    c = f.applied("#IV7", 1)
    assert c.root().natural() == "F#"
    assert c.type() == "7"


def test_applied_string_string():
    f = Field("C", ScaleType.Major)
    # V7 / IIm = A7
    c = f.applied("V7", "IIm")
    assert c.root().natural() == "A"
    assert c.type() == "7"
    # V7 / V = D7
    c = f.applied("V7", "V")
    assert c.root().natural() == "D"
    assert c.type() == "7"


def test_applied_numeric():
    f = Field("C", ScaleType.Major)
    # applied(5, 2) = seventh at V of D major = A7
    c = f.applied(5, 2)
    assert c.root().natural() == "A"
    assert c.type() == "7"
    # applied(5, 5) = D7
    c = f.applied(5, 5)
    assert c.root().natural() == "D"
    assert c.type() == "7"


def test_applied_out_of_range():
    f = Field("C", ScaleType.Major)
    with pytest.raises(IndexError):
        f.applied("V7", 0)
    with pytest.raises(IndexError):
        f.applied("V7", 8)
    with pytest.raises(IndexError):
        f.applied(5, 0)


def test_applied_invalid_roman():
    f = Field("C", ScaleType.Major)
    with pytest.raises(ValueError):
        f.applied("X7", 2)
    with pytest.raises(ValueError):
        f.applied("", 2)


# ---------------------------------------------------------------------------
# Harmonic function and role
# ---------------------------------------------------------------------------

def test_function_by_degree():
    f = Field("C", ScaleType.Major)

    assert f.function(1) == HarmonicFunction.Tonic
    assert f.function(2) == HarmonicFunction.Subdominant
    assert f.function(3) == HarmonicFunction.Tonic
    assert f.function(4) == HarmonicFunction.Subdominant
    assert f.function(5) == HarmonicFunction.Dominant
    assert f.function(6) == HarmonicFunction.Tonic
    assert f.function(7) == HarmonicFunction.Dominant


def test_function_name_and_short():
    assert HarmonicFunction.Tonic.name == "Tonic"
    assert HarmonicFunction.Subdominant.name == "Subdominant"
    assert HarmonicFunction.Dominant.name == "Dominant"

    assert HarmonicFunction.Tonic.short == "T"
    assert HarmonicFunction.Subdominant.short == "S"
    assert HarmonicFunction.Dominant.short == "D"


def test_role_by_degree():
    f = Field("C", ScaleType.Major)

    assert f.role(1) == "primary"
    assert f.role(2) == "relative of IV"
    assert f.role(3) == "transitive"
    assert f.role(4) == "primary"
    assert f.role(5) == "primary"
    assert f.role(6) == "relative of I"
    assert f.role(7) == "relative of V"


def test_function_by_chord_name():
    f = Field("C", ScaleType.Major)

    # FM is degree IV → Subdominant
    result = f.function("FM")
    assert result == HarmonicFunction.Subdominant

    # Am is degree VI → Tonic
    result = f.function("Am")
    assert result == HarmonicFunction.Tonic

    # GM is degree V → Dominant
    result = f.function("GM")
    assert result == HarmonicFunction.Dominant


def test_function_by_chord_object():
    f = Field("C", ScaleType.Major)
    chord = Chord("Dm")

    result = f.function(chord)
    assert result == HarmonicFunction.Subdominant


def test_function_returns_none_for_non_field_chord():
    f = Field("C", ScaleType.Major)

    # F#M does not belong to C major
    result = f.function("F#M")
    assert result is None


def test_role_by_chord_name():
    f = Field("C", ScaleType.Major)

    result = f.role("Am")
    assert result == "relative of I"

    result = f.role("FM")
    assert result == "primary"

    # Non-field chord
    result = f.role("F#M")
    assert result is None


def test_function_out_of_range():
    f = Field("C", ScaleType.Major)
    with pytest.raises(IndexError):
        f.function(0)
    with pytest.raises(IndexError):
        f.function(8)
    with pytest.raises(IndexError):
        f.role(0)


def test_function_natural_minor():
    f = Field("A", ScaleType.NaturalMinor)

    assert f.function(1) == HarmonicFunction.Tonic
    assert f.function(3) == HarmonicFunction.Tonic
    assert f.function(5) == HarmonicFunction.Dominant
    assert f.function(6) == HarmonicFunction.Subdominant
    assert f.function(7) == HarmonicFunction.Dominant


# ---------------------------------------------------------------------------
# Circle of fifths navigation
# ---------------------------------------------------------------------------


def test_field_signature():
    assert Field("C", "major").signature() == 0
    assert Field("G", "major").signature() == 1
    assert Field("D", "major").signature() == 2
    assert Field("F", "major").signature() == -1
    assert Field("Bb", "major").signature() == -2
    assert Field("A", "natural minor").signature() == 0


def test_field_relative_major():
    rel = Field("C", "major").relative()
    assert rel.tonic().natural() == "A"
    assert len(rel.chords()) == 7


def test_field_relative_minor():
    rel = Field("A", "natural minor").relative()
    assert rel.tonic().natural() == "C"
    assert len(rel.chords()) == 7


def test_field_relative_g_major():
    rel = Field("G", "major").relative()
    assert rel.tonic().natural() == "E"


def test_field_parallel_major():
    f = Field("C", "major")
    par = f.parallel()
    assert par.tonic().natural() == "C"
    assert len(par.chords()) == 7


def test_field_parallel_minor():
    par = Field("C", "natural minor").parallel()
    assert par.tonic().natural() == "C"


def test_field_neighbors_c_major():
    sub, dom = Field("C", "major").neighbors()
    assert sub.tonic().natural() == "F"
    assert dom.tonic().natural() == "G"
    assert len(sub.chords()) == 7
    assert len(dom.chords()) == 7


def test_field_neighbors_a_minor():
    sub, dom = Field("A", "natural minor").neighbors()
    assert sub.tonic().natural() == "D"
    assert dom.tonic().natural() == "E"

File: tests/cpp/test_field.cpp
// Gingo — Field module tests
// SPDX-License-Identifier: MIT

#include <catch2/catch_test_macros.hpp>
#include <gingo/field.hpp>

using namespace gingo;

TEST_CASE("Field construction", "[field]") {
    Field f("C", ScaleType::Major);
    REQUIRE(f.tonic().natural() == "C");
    REQUIRE(f.size() == 7);
}

TEST_CASE("Field from string", "[field]") {
    Field f("C", "major");
    REQUIRE(f.size() == 7);
}

TEST_CASE("Field triads count", "[field]") {
    Field f("C", ScaleType::Major);
    auto chords = f.chords();
    REQUIRE(chords.size() == 7);
}

TEST_CASE("Field sevenths count", "[field]") {
    Field f("C", ScaleType::Major);
    auto chords = f.sevenths();
    REQUIRE(chords.size() == 7);
}

TEST_CASE("Field chord at degree", "[field]") {
    Field f("C", ScaleType::Major);
    auto c1 = f.chord(1);
    REQUIRE(c1.root().natural() == "C");
}

TEST_CASE("Field chord at invalid degree throws", "[field]") {
    Field f("C", ScaleType::Major);
    REQUIRE_THROWS(f.chord(0));
    REQUIRE_THROWS(f.chord(8));
}

// ---------------------------------------------------------------------------
// Applied chords (tonicization)
// ---------------------------------------------------------------------------

TEST_CASE("Field applied string + int: secondary dominants", "[field]") {
    Field f("C", ScaleType::Major);

    // V7/II  → V7 of D major = A7
    auto c = f.applied("V7", 2);
    REQUIRE(c.root().natural() == "A");
    REQUIRE(c.type() == "7");

    // V7/IV  → V7 of F major = C7
    c = f.applied("V7", 4);
    REQUIRE(c.root().natural() == "C");
    REQUIRE(c.type() == "7");

    // V7/V   → V7 of G major = D7
    c = f.applied("V7", 5);
    REQUIRE(c.root().natural() == "D");
    REQUIRE(c.type() == "7");
}

TEST_CASE("Field applied string + int: complex quality", "[field]") {
    Field f("C", ScaleType::Major);

    // IIm7(b5)/V  → II of G major = A, quality m7(b5) → Am7(b5)
    auto c = f.applied("IIm7(b5)", 5);
    REQUIRE(c.root().natural() == "A");
    REQUIRE(c.type() == "m7(b5)");
}

TEST_CASE("Field applied string + int: no quality returns triad", "[field]") {
    Field f("C", ScaleType::Major);

    // V/II  → triad at V of D major = AM
    auto c = f.applied("V", 2);
    REQUIRE(c.root().natural() == "A");
}

TEST_CASE("Field applied with accidental", "[field]") {
    Field f("C", ScaleType::Major);

    // bVII7/V  → VII of G major = F#, flatten → F, quality 7 → F7
    auto c = f.applied("bVII7", 5);
    REQUIRE(c.root().natural() == "F");
    REQUIRE(c.type() == "7");

    // #IV7/I  → IV of C major = F, sharpen → F#, quality 7 → F#7
    c = f.applied("#IV7", 1);
    REQUIRE(c.root().natural() == "F#");
    REQUIRE(c.type() == "7");
}

TEST_CASE("Field applied string + string: Roman numeral target", "[field]") {
    Field f("C", ScaleType::Major);

    // V7 / IIm  → V7 of degree II = A7
    auto c = f.applied("V7", "IIm");
    REQUIRE(c.root().natural() == "A");
    REQUIRE(c.type() == "7");

    // V7 / V  → D7
    c = f.applied("V7", "V");
    REQUIRE(c.root().natural() == "D");
    REQUIRE(c.type() == "7");

    // IIm / IV  → IIm of degree IV = triad at II of F major = Gm
    c = f.applied("IIm", "IV");
    REQUIRE(c.root().natural() == "G");
}

TEST_CASE("Field applied numeric shorthand", "[field]") {
    Field f("C", ScaleType::Major);

    // applied(5, 2) = seventh at V of D major = A7
    auto c = f.applied(5, 2);
    REQUIRE(c.root().natural() == "A");
    REQUIRE(c.type() == "7");

    // applied(5, 5) = seventh at V of G major = D7
    c = f.applied(5, 5);
    REQUIRE(c.root().natural() == "D");
    REQUIRE(c.type() == "7");
}

TEST_CASE("Field applied out of range", "[field]") {
    Field f("C", ScaleType::Major);
    REQUIRE_THROWS(f.applied("V7", 0));
    REQUIRE_THROWS(f.applied("V7", 8));
    REQUIRE_THROWS(f.applied(5, 0));
}

TEST_CASE("Field applied invalid Roman numeral", "[field]") {
    Field f("C", ScaleType::Major);
    REQUIRE_THROWS(f.applied("X7", 2));
    REQUIRE_THROWS(f.applied("", 2));
}

// ---------------------------------------------------------------------------
// Harmonic function and role
// ---------------------------------------------------------------------------

TEST_CASE("Field function by degree: C major", "[field]") {
    Field f("C", ScaleType::Major);

    REQUIRE(f.function(1) == HarmonicFunction::Tonic);
    REQUIRE(f.function(2) == HarmonicFunction::Subdominant);
    REQUIRE(f.function(3) == HarmonicFunction::Tonic);
    REQUIRE(f.function(4) == HarmonicFunction::Subdominant);
    REQUIRE(f.function(5) == HarmonicFunction::Dominant);
    REQUIRE(f.function(6) == HarmonicFunction::Tonic);
    REQUIRE(f.function(7) == HarmonicFunction::Dominant);
}

TEST_CASE("Field function name and short", "[field]") {
    REQUIRE(harmonic_function_name(HarmonicFunction::Tonic) == "Tonic");
    REQUIRE(harmonic_function_name(HarmonicFunction::Subdominant) == "Subdominant");
    REQUIRE(harmonic_function_name(HarmonicFunction::Dominant) == "Dominant");

    REQUIRE(harmonic_function_short(HarmonicFunction::Tonic) == "T");
    REQUIRE(harmonic_function_short(HarmonicFunction::Subdominant) == "S");
    REQUIRE(harmonic_function_short(HarmonicFunction::Dominant) == "D");
}

TEST_CASE("Field role by degree: C major", "[field]") {
    Field f("C", ScaleType::Major);

    REQUIRE(f.role(1) == "primary");
    REQUIRE(f.role(2) == "relative of IV");
    REQUIRE(f.role(3) == "transitive");
    REQUIRE(f.role(4) == "primary");
    REQUIRE(f.role(5) == "primary");
    REQUIRE(f.role(6) == "relative of I");
    REQUIRE(f.role(7) == "relative of V");
}

TEST_CASE("Field function by chord name", "[field]") {
    Field f("C", ScaleType::Major);

    // FM is degree IV → Subdominant
    auto result = f.function("FM");
    REQUIRE(result.has_value());
    REQUIRE(*result == HarmonicFunction::Subdominant);

    // Am is degree VI → Tonic
    result = f.function("Am");
    REQUIRE(result.has_value());
    REQUIRE(*result == HarmonicFunction::Tonic);

    // GM is degree V → Dominant
    result = f.function("GM");
    REQUIRE(result.has_value());
    REQUIRE(*result == HarmonicFunction::Dominant);
}

TEST_CASE("Field function by Chord object", "[field]") {
    Field f("C", ScaleType::Major);
    Chord chord("Dm");

    auto result = f.function(chord);
    REQUIRE(result.has_value());
    REQUIRE(*result == HarmonicFunction::Subdominant);
}

TEST_CASE("Field function returns nullopt for non-field chord", "[field]") {
    Field f("C", ScaleType::Major);

    // F#M does not belong to C major
    auto result = f.function("F#M");
    REQUIRE(!result.has_value());
}

TEST_CASE("Field role by chord name", "[field]") {
    Field f("C", ScaleType::Major);

    auto result = f.role("Am");
    REQUIRE(result.has_value());
    REQUIRE(*result == "relative of I");

    result = f.role("FM");
    REQUIRE(result.has_value());
    REQUIRE(*result == "primary");

    // Non-field chord
    result = f.role("F#M");
    REQUIRE(!result.has_value());
}

TEST_CASE("Field function out of range", "[field]") {
    Field f("C", ScaleType::Major);
    REQUIRE_THROWS(f.function(0));
    REQUIRE_THROWS(f.function(8));
    REQUIRE_THROWS(f.role(0));
}

TEST_CASE("Field function natural minor", "[field]") {
    Field f("A", ScaleType::NaturalMinor);

    // i → T, ii → S, III → T, iv → S, v → D, VI → S, VII → D
    REQUIRE(f.function(1) == HarmonicFunction::Tonic);
    REQUIRE(f.function(3) == HarmonicFunction::Tonic);
    REQUIRE(f.function(5) == HarmonicFunction::Dominant);
    REQUIRE(f.function(6) == HarmonicFunction::Subdominant);
    REQUIRE(f.function(7) == HarmonicFunction::Dominant);
}

// ---------------------------------------------------------------------------
// Circle of fifths navigation
// ---------------------------------------------------------------------------

TEST_CASE("Field signature delegates to scale", "[field][fifths]") {
    REQUIRE(Field("C", "major").signature() == 0);
    REQUIRE(Field("G", "major").signature() == 1);
    REQUIRE(Field("D", "major").signature() == 2);
    REQUIRE(Field("F", "major").signature() == -1);
    REQUIRE(Field("Bb", "major").signature() == -2);
    REQUIRE(Field("A", "natural minor").signature() == 0);
}

TEST_CASE("Field relative major to minor", "[field][fifths]") {
    Field f("C", "major");
    Field rel = f.relative();
    REQUIRE(rel.tonic().natural() == "A");
    REQUIRE(rel.chords().size() == 7);
}

TEST_CASE("Field relative minor to major", "[field][fifths]") {
    Field f("A", "natural minor");
    Field rel = f.relative();
    REQUIRE(rel.tonic().natural() == "C");
    REQUIRE(rel.chords().size() == 7);
}

TEST_CASE("Field relative G major", "[field][fifths]") {
    Field rel = Field("G", "major").relative();
    REQUIRE(rel.tonic().natural() == "E");
}

TEST_CASE("Field parallel major to minor", "[field][fifths]") {
    Field f("C", "major");
    Field par = f.parallel();
    REQUIRE(par.tonic().natural() == "C");
    REQUIRE(par.chords().size() == 7);
    // Parallel should produce different chords
    REQUIRE(f.chord(3).name() != par.chord(3).name());
}

TEST_CASE("Field parallel minor to major", "[field][fifths]") {
    Field par = Field("C", "natural minor").parallel();
    REQUIRE(par.tonic().natural() == "C");
    // C major chord I = CM
    REQUIRE(par.chord(1).root().natural() == "C");
}

TEST_CASE("Field neighbors C major", "[field][fifths]") {
    auto [sub, dom] = Field("C", "major").neighbors();
    REQUIRE(sub.tonic().natural() == "F");
    REQUIRE(dom.tonic().natural() == "G");
    REQUIRE(sub.chords().size() == 7);
    REQUIRE(dom.chords().size() == 7);
}

TEST_CASE("Field neighbors A minor", "[field][fifths]") {
    auto [sub, dom] = Field("A", "natural minor").neighbors();
    REQUIRE(sub.tonic().natural() == "D");
    REQUIRE(dom.tonic().natural() == "E");
}
